### PR TITLE
Opaque equality for types without `operator==`

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/opaque_type.h [hunk 1]
```diff
@@ -79,7 +82,11 @@ class Opaque {
 
 template <typename T>
 bool OpaqueEquals<T>::operator()(const Opaque& a, const Opaque& b) const {
-  return a.as<T>() == b.as<T>();
+  if constexpr (dynamic_type::opcheck<T> == dynamic_type::opcheck<T>) {
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply csrc/opaque_type.h:85]: Yes. It returns false at least for the test case. Just feel the format is confusing. Thanks for the detailed explain. Can we add what you wrote as comment to the code:

> opcheck is specifically designed for the existence of operators, so the operators on opcheck behave differently from normal C++. For example, if you assign one opcheck to another opcheck, it will return a bool telling you whether this asignment is valid for the underlying type, instead of returning a reference of the lhs opcheck.

There are some comments on `opcheck` but seems out of date. e.g. `// For more examples, see test_dynamic_type.cpp namespace opcheck_tests`

File: csrc/opaque_type.h [hunk 2]
```diff
@@ -79,7 +82,11 @@ class Opaque {
...
+    return a.as<T>() == b.as<T>();
+  } else {
+    return std::memcmp(&a.as<T>(), &b.as<T>(), sizeof(T)) == 0;
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/opaque_type.h:88]: You are absolutely right that this implementation will not always correctly compare objects, and your example totally make senses. The problem here is, because C++ is a statically typed language, there is no general way to compare opaque values. So we can only have an approximation of `operator==`, which does not always work. For example, `1 == 1.0`, but `Opaque(1) != Opaque(1.0)`.  This PR adds another approximation, and just like any other approximation, it is not always correct. The only guarantee we can have is: if you are comparing two opaque objects with identical underlying type, and that underlying type defined `operator==`, then you will get the desired behavior.

The problem of bitwise comparison is not just in the example you mentioned. For example, if I have
```C++
struct A {
  int32_t i;
  double d;
};
```
Then usually C++ compiler will allocate memory like this
```
[======== i (32bits) ========][======== empty (32bits) ========][======== d (64bits) ========]
```
And the empty bits will be filled with undefined values, and bitwise compare will fail.

I don't think we will use the bitwise comparison today. This path exist only because, otherwise, this code will fail to compile if we want to create an opaque object a concrete object without `operator==`.

For the future, if we encounter a case where we know it will work (for example, it is a packed POD), we can omit the `operator==`.

[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/opaque_type.h:88]: @zasdfgbnm Can you please add this as a code comment?

**Review by liqiangxl — APPROVED** [review]
Thanks for clarify the questions.

File: csrc/opaque_type.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/opaque_type.h:89]: comment added

File: csrc/opaque_type.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/opaque_type.h:114]: added

File: lib/dynamic_type/benchmark/knn.cpp [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 #include <benchmark/benchmark.h>
 
-#include <dynamic_type.h>
+#include <dynamic_type/dynamic_type.h>
```
Comment (zasdfgbnm) [inline lib/dynamic_type/benchmark/knn.cpp:11]: unrelated build fix
